pkg/traces/remotewriteexporter/exporter.go

```diff
@@ -192,18 +198,81 @@ func (e *remoteWriteExporter) handleHistogramDataPoints(app storage.Appender, na
 			cumulativeCount += dataPoint.BucketCounts()[ix]
 			boundStr := strconv.FormatFloat(eb, 'f', -1, 64)
 			bucketLabels := e.createLabelSet(name, bucketSuffix, dataPoint.Attributes(), labels.Labels{{Name: leStr, Value: boundStr}})
-			if _, err := app.Append(0, bucketLabels, ts, float64(cumulativeCount)); err != nil {
-				return err
-			}
+			e.appendDatapointForSeries(bucketLabels, ts, float64(cumulativeCount))
 		}
 		// add le=+Inf bucket
 		cumulativeCount += dataPoint.BucketCounts()[len(dataPoint.BucketCounts())-1]
 		infBucketLabels := e.createLabelSet(name, bucketSuffix, dataPoint.Attributes(), labels.Labels{{Name: leStr, Value: infBucket}})
-		if _, err := app.Append(0, infBucketLabels, ts, float64(cumulativeCount)); err != nil {
-			return err
+		e.appendDatapointForSeries(infBucketLabels, ts, float64(cumulativeCount))
+	}
+}
+
+func (e *remoteWriteExporter) appendDatapointForSeries(l labels.Labels, ts int64, v float64) {
+	e.mtx.Lock()
+	defer e.mtx.Unlock()
+
+	series := l.Hash()
+	if lastDatapoint, ok := e.seriesMap[series]; ok {
+		if lastDatapoint.ts >= ts {
+			return
+		}
+		lastDatapoint.ts = ts
+		lastDatapoint.v = v
+		return
+	}
+
+	e.seriesMap[series] = &datapoint{l: l, ts: ts, v: v}
+}
+
+func (e *remoteWriteExporter) appenderLoop() {
+	loopInterval := 2 * time.Second
+	t := time.NewTicker(loopInterval)
+
+	for {
+		select {
+		case <-t.C:
+			e.mtx.Lock()
+			inst, err := e.manager.GetInstance(e.promInstance)
+			if err != nil {
+				level.Error(e.logger).Log("msg", "failed to get prom instance", "err", err)
+				continue
+			}
+			appender := inst.Appender(context.Background())
+
+			now := time.Now().UnixMilli()
+			var staleSeries []uint64
+			for _, dp := range e.seriesMap {
+				// If the datapoint hasn't been updated since the last loop, don't append it
+				if dp.ts < e.lastFlush {
+					// If the datapoint is older than now - staleTime, it is stale and should be removed.
+					if now-dp.ts > e.staleTime {
+						staleSeries = append(staleSeries, dp.l.Hash())
```

Comment:
Could you do the `delete(e.seriesMap, hash)` here directly?

Reply:
Yup, thanks :)